cartofacade/queue.go

```diff
@@ -0,0 +1,194 @@
+// Package cartofacade contains the api to call into CGO
+package cartofacade
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	cgoApi "github.com/viamrobotics/viam-cartographer/cartofacade/internal/capi"
+)
+
+// WorkType defines the grpc call that is being made.
+type WorkType int64
+
+const (
+	// Initialize can be used to represent the viam_carto_init call in c.
+	Initialize WorkType = iota
+	// Start can be used to represent the viam_carto_start call in c.
+	Start
+	// Stop can be used to represent the viam_carto_stop call in c.
+	Stop
+	// Terminate can be used to represent the viam_carto_terminate in c.
+	Terminate
+	// AddSensorReading can be used to represent the viam_carto_add_sensor_reading in c.
+	AddSensorReading
+	// Position can be used to represent the viam_carto_get_position call in c.
+	Position
+	// InternalState can be used to represent the viam_carto_get_internal_state call in c.
+	InternalState
+	// PointCloudMap can be used to represent the viam_carto_get_point_cloud_map in c.
+	PointCloudMap
+)
+
+// InputType defines the type being provided as input to the work.
+type InputType int64
+
+const (
+	// Sensor represents a sensor name input into c funcs.
+	Sensor InputType = iota
+	// Reading represents a lidar reading input into c funcs.
+	Reading
+	// Timestamp represents the timestamp input into c funcs.
+	Timestamp
+)
+
+// ResultType defines the type being provided as input to the work.
+type ResultType int64
+
+const (
+	// ByteSlice represents the go type []Byte.
+	ByteSlice ResultType = iota
+	// PositionResponse represents the custom struct defined for getPosition responses.
+	PositionResponse
+	// CartoType represents the custom struct defined for Carto structs.
+	CartoType
+	// Nil represents the go type nil.
+	Nil
+	// Error represents the go type error.
+	Error
+)
+
+// Response defines the result of one piece of work that can be put on the result channel.
+type Response struct {
+	Result     interface{}
+	ResultType ResultType
+}
+
+// Queue represents a queue to consume work from and enforce one call into C at a time.
+type Queue struct {
+	WorkChannel     chan WorkItem
+	CartoLib        *cgoApi.CartoLibInterface
+	Carto           cgoApi.CartoInterface
+	CartoConfig     cgoApi.CartoConfig
+	CartoAlgoConfig cgoApi.CartoAlgoConfig
+}
+
+// WorkItemInterface defines one piece of work that can be put on the queue.
+type WorkItemInterface interface {
+	DoWork(q *Queue) (interface{}, ResultType, error)
+}
+
+// WorkItem defines one piece of work that can be put on the queue.
+type WorkItem struct {
+	Result   chan Response
+	workType WorkType
+	inputs   map[InputType]interface{}
+}
+
+// NewQueue instantiates the Queue struct.
+func NewQueue(cartoLib cgoApi.CartoLibInterface, cartoCfg cgoApi.CartoConfig, cartoAlgoCfg cgoApi.CartoAlgoConfig) Queue {
```

Comment:
[opt] lets try to make it so that none of the parameters to this function live in the cgoApi package. We would like the callers of cartofacade to not need to not need to know that the CGO api exists.